### PR TITLE
New Topic Page: Original Request Context

### PR DESCRIPTION
**Pull request by alexfornuto**
## Summary

New Topic doc describing how `X-Pomerium-Jwt-Assertion-For` preserves the original request JWT for the service account requests to secondary services.

## Related issues

Closes https://github.com/pomerium/internal/issues/563

## Checklist

- [x] reference any related issues
- [x] updated docs
- [ ] updated unit tests
- [ ] updated UPGRADING.md
- [x] add appropriate tag (`improvement` / `bug` / etc)
- [x] ready for review

### CODE REVIEW

**Comment by alexfornuto** [comment]
Not ready for full review, but CC @travisgroth to confirm I'm barking down the right path.

**Comment by travisgroth** [comment]
> Not ready for full review, but CC @travisgroth to confirm I'm barking down the right path.

👍 looks good so far

File: docs/docs/topics/original-request-context.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+description: This article describes how the original user context is passed secondary requests.
+---
+
+# Original User Context
+
+::: tip
+This article describes a use case available to [Pomerium Enterprise](/enterprise/about.md) customers.
+:::
+
+In enterprise environments where multiple services protected by Pomerium communicate with each other using a [service account](/enterprise/concepts.md#service-accounts), there are scenarios where the original user context must be preserved. This article describes how this is accomplished with the `X-Pomerium-Jwt-Assertion-For` header.
```
Comment (wasaga) [inline docs/docs/topics/original-request-context.md:12]: we probably need have a page (and reference from here) how you can protect internal api resources with service accounts.

File: docs/docs/topics/original-request-context.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+```
+
+- **App** is a user-facing application. Users connect to it through Pomerium.
+- **API** is also accessed through it's Pomerium Route, but is only accessible by the **App**, using a [service account](/enterprise/concepts.md#service-accounts) to authenticate.
+- The **API** service needs to know the user making the request to **App** in order to return the needed 
+
+Both Routes include [`pass_identity_headers`](/reference.md#pass-identity-headers), which provides (at minimum) the `X-Pomerium-Jwt-Assertion` header to the downstream application.
+
+- When a user connects to **App** to perform an action, `X-Pomerium-Jwt-Assertion` provides the service with the Json Web Token (**JWT**) for the user.
+- When **App** sends a request to **API**, `X-Pomerium-Jwt-Assertion` is provided for the service account.
```
Comment (wasaga) [inline docs/docs/topics/original-request-context.md:47]: it sounds like it's automatic while it's not. a calling application has to copy `x-pomerium-jwt-assertion` from original request

File: docs/docs/topics/original-request-context.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+- Pomerium copies the original user context, the JWT from the user, to `X-Pomerium-Jwt-Assertion-For` and includes it in the request to **API**
+
+Now the **API** can service can read `X-Pomerium-Jwt-Assertion-For` as needed to determine the proper response to send.
```
Comment (wasaga) [inline docs/docs/topics/original-request-context.md:50]: also this process can be repeated many times (second time though with x-pomerium-jwt-assertion-for

**Review by wasaga — DISMISSED** [review]

File: docs/docs/topics/original-request-context.md (hunk 3 above)
Comment (alexfornuto) [reply docs/docs/topics/original-request-context.md:50]: Can you please extrapolate? I don't follow how it's different the second time.

File: docs/docs/topics/original-request-context.md (hunk 2 above)
Comment (alexfornuto) [reply docs/docs/topics/original-request-context.md:47]: In this step, isn't `x-pomerium-jwt-assersion` the JWT of the service account, not the user? 

If so, should I specify that the **App** service should already be configured by the developer to use the Service Account JWT?

If not, can you please clarify?

File: docs/docs/topics/original-request-context.md (hunk 1 above)
Comment (alexfornuto) [reply docs/docs/topics/original-request-context.md:12]: Started in 4521af68cef3876712ee91e81879fc2ad0e92471

File: docs/docs/topics/original-request-context.md (hunk 2 above)
Comment (alexfornuto) [reply docs/docs/topics/original-request-context.md:47]: Or, are you saying it's at this point that the app has to take the step of copying what it received as `x-pomerium-jwt-assertion` and send it as part of it's request to Pomerium as `x-pomerium-assertion-for`?

File: docs/docs/topics/original-request-context.md (hunk 3 above)
Comment (alexfornuto) [reply docs/docs/topics/original-request-context.md:50]: Spoke with @travisgroth, and got the understanding I need to adjust.

File: docs/.vuepress/config.js [hunk 4]
```diff
@@ -5,6 +5,7 @@ module.exports = {
     "Pomerium is a beyond-corp inspired, zero trust, open source identity-aware access proxy.",
   plugins: [
     "vuepress-plugin-element-tabs",
+    'vuepress-plugin-mermaidjs',
```
[the remark this answers is not in the document]
Comment (alexfornuto) [reply docs/.vuepress/config.js:8]: it's a coming...

File: docs/docs/topics/original-request-context.md (hunk 3 above)
Comment (alexfornuto) [reply docs/docs/topics/original-request-context.md]: Discussed [here](https://github.com/pomerium/pomerium/pull/2569/files#diff-f61107ecd4bbaacb87080c7226ef5704adfe1d83794f53ee4937ee610e72b6eaR25) and [here](https://github.com/pomerium/pomerium/pull/2569/files#diff-f61107ecd4bbaacb87080c7226ef5704adfe1d83794f53ee4937ee610e72b6eaR97-R122)

File: docs/docs/topics/original-request-context.md (hunk 2 above)
Comment (alexfornuto) [reply docs/docs/topics/original-request-context.md]: [Updated](https://github.com/pomerium/pomerium/pull/2569/files#diff-f61107ecd4bbaacb87080c7226ef5704adfe1d83794f53ee4937ee610e72b6eaR90)

**Review by travisgroth — COMMENTED** [review]
LGTM

File: docs/enterprise/service-accounts.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+1. From the main menu, select **Service Accounts** under **CONFIGURE**. Click the **+ ADD SERVICE ACCOUNT** button:
+
+   ![An empty Service Accounts page](./img/add-service-account.png)
+
+1. Service accounts can be unique and exist only for Pomerium, or impersonate directory users from your IdP.
+
+   ::::: tabs
+   :::: tab Unique
+   Give the user a unique ID. Consider referencing the Namespace you're creating it under, for easier reference later. Optionally set an expiration date:
```
Comment (desimone) [inline docs/enterprise/service-accounts.md:25]: We should note that the User ID value here would correspond to the `User` criteria on the policy screen.

E.G. If they want to say 

```
allow:
  or:
    - user:
        is: "whatever unique user id from above"
```

**Review by desimone — CHANGES_REQUESTED** [review]
Just one suggestion to make clear what the User ID corresponds to in terms of policy.

File: docs/enterprise/service-accounts.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+   ::::
+   :::::
+
+1. After you click **Submit**, the modal presents the Java Web Token (**JWT**) for the service account. Temporarily save it somewhere secure, as you will not be able to view it again:
+
+   ![Service Account Added](./img/service-account-jwt.png)
+
+   This JWT must be added to your application configuration to enable direct communication.
+
+1. Edit or create policies to give the service account access to the internal service.
```
Comment (desimone) [inline docs/enterprise/service-accounts.md:46]: Maybe even drive it home here too that the policy would use the User criteria matching on the device id.

File: docs/enterprise/service-accounts.md (hunk 5 above)
Comment (alexfornuto) [reply docs/enterprise/service-accounts.md:25]: (Hopefully) addressed in b04b67f146194f8ae58ed97f29ba88b50bda9c73.

File: docs/enterprise/service-accounts.md (hunk 6 above)
Comment (alexfornuto) [reply docs/enterprise/service-accounts.md]: (Hopefully) addressed in b04b67f146194f8ae58ed97f29ba88b50bda9c73.

**Review by desimone — APPROVED** [review]

**Pull request merged**